contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -0,0 +1,700 @@
+	    (uint8_t *)key->value, WRAPPING_KEY_LEN);
+	pw_free(key);
+	if (ret) {
+		pam_syslog(pamh, LOG_ERR, "change_key failed: %d", ret);
+		nvlist_free(nvlist);
+		zfs_close(ds);
+		return (-1);
+	}
+	return (0);
+}
+
+static int
+decrypt_mount(pam_handle_t *pamh, const char *ds_name,
+    const char *passphrase)
+{
+	zfs_handle_t *ds = zfs_open(libzfs, ds_name, ZFS_TYPE_FILESYSTEM);
+	if (ds == NULL) {
+		pam_syslog(pamh, LOG_ERR, "dataset %s not found", ds_name);
+		return (-1);
+	}
+	password *key = prepare_passphrase(pamh, ds, passphrase, NULL);
+	if (key == NULL) {
+		zfs_close(ds);
+		return (-1);
+	}
+	int ret = lzc_load_key(ds_name, B_FALSE, (uint8_t *)key->value,
+	    WRAPPING_KEY_LEN);
+	pw_free(key);
+	if (ret) {
+		pam_syslog(pamh, LOG_ERR, "load_key failed: %d", ret);
+		zfs_close(ds);
+		return (-1);
+	}
+	ret = zfs_mount(ds, NULL, 0);
+	if (ret) {
+		pam_syslog(pamh, LOG_ERR, "mount failed: %d", ret);
+		zfs_close(ds);
+		return (-1);
+	}
+	return (0);
+}
+
+static int
+unmount_unload(pam_handle_t *pamh, const char *ds_name)
+{
+	zfs_handle_t *ds = zfs_open(libzfs, ds_name, ZFS_TYPE_FILESYSTEM);
+	if (ds == NULL) {
+		pam_syslog(pamh, LOG_ERR, "dataset %s not found", ds_name);
+		return (-1);
+	}
+	int ret = zfs_unmount(ds, NULL, 0);
+	if (ret) {
+		pam_syslog(pamh, LOG_ERR, "zfs_unmount failed with: %d", ret);
+		zfs_close(ds);
+		return (-1);
+	}
+
+	ret = lzc_unload_key(ds_name);
+	if (ret) {
+		pam_syslog(pamh, LOG_ERR, "unload_key failed with: %d", ret);
+		zfs_close(ds);
+		return (-1);
+	}
+	zfs_close(ds);
+	return (0);
+}
+
+static void
+sigchild_default(struct sigaction *oldact)
+{
+	struct sigaction newact;
+	newact.sa_handler = SIG_DFL;
+	newact.sa_flags = 0;
+	sigfillset(&newact.sa_mask);
+	sigaction(SIGCHLD, &newact, oldact);
+}
+
+static void
+sigchild_default_reset(struct sigaction *oldact)
+{
+	sigaction(SIGCHLD, oldact, NULL);
+}
+
+typedef struct {
+	char *homes_prefix;
+	uid_t uid;
+	const char *username;
+} zfs_key_config;
+
+static int
+zfs_key_config_load(pam_handle_t *pamh, zfs_key_config *config,
+    int argc, const char **argv)
+{
+	config->homes_prefix = strdup("rpool/home");
+	const char *name;
+	if (pam_get_user(pamh, &name, NULL) != PAM_SUCCESS) {
+		pam_syslog(pamh, LOG_ERR,
+		    "couldn't get username from PAM stack");
+		free(config->homes_prefix);
+		return (-1);
+	}
+	struct passwd *entry = getpwnam(name);
+	config->uid = entry->pw_uid;
+	config->username = name;
+	for (int c = 0; c < argc; c++) {
+		if (strncmp(argv[c], "homes=", 6) == 0) {
+			free(config->homes_prefix);
+			config->homes_prefix = strdup(argv[c] + 6);
+		}
+	}
+	return (0);
+}
+
+static void
+zfs_key_config_free(zfs_key_config *config)
+{
+	free(config->homes_prefix);
+}
+
+static char *
+zfs_key_config_get_dataset(zfs_key_config *config)
+{
+	size_t len = ZFS_MAX_DATASET_NAME_LEN;
+	char *ret = malloc(len + 1);
+	ret[0] = 0;
+	strncat(ret, config->homes_prefix, len);
+	strncat(ret, "/", len);
```

Comment:
Improper use of the 3rd argument. You must use `strlcat`, or reduce the len arg.